src/serialize_packed.rs

```diff
@@ -357,62 +326,48 @@ impl <'a, W : BufferedOutputStream> OutputStream for PackedOutputStream<'a, W> {
-                        ::std::ptr::copy_nonoverlapping(src, out, count);
+                    let count: usize = ptr_sub(in_ptr, run_start);
+                    *buf.get_unchecked_mut(buf_idx) = (count / 8) as u8;
+                    buf_idx += 1;
 
-                        out = out.offset(count as isize);
-                    } else {
-                        //# Input overruns the output buffer. We'll give it
-                        //# to the output stream in one chunk and let it
-                        //# decide what to do.
-                        try!(self.inner.write_ptr(buffer_begin, ptr_sub(out, buffer_begin)));
-
-                        {
-                            let buf = ::std::slice::from_raw_parts::<u8>(run_start, count);
-                            try!(self.inner.write(buf));
-                        }
-
-                        let (out1, buffer_end1) = self.inner.get_write_buffer();
-                        out = out1; buffer_end = buffer_end1;
-                        buffer_begin = out;
-                    }
+                    try!(self.inner.write_all(&buf[..buf_idx]));
+                    buf_idx = 0;
+                    try!(self.inner.write_all(slice::from_raw_parts::<u8>(run_start, count)));
                 }
             }
 
-            try!(self.inner.write_ptr(buffer_begin, ptr_sub(out, buffer_begin)));
-            Ok(())
+            try!(self.inner.write_all(&buf[..buf_idx]));
+            Ok(in_buf.len())
         }
     }
 
-   fn flush(&mut self) -> ::std::io::Result<()> { self.inner.flush() }
+   fn flush(&mut self) -> io::Result<()> { self.inner.flush() }
 }
 
-pub fn write_packed_message<T: BufferedOutputStream, U: MessageBuilder>(
-    output : &mut T, message : &mut U) -> ::std::io::Result<()> {
-    let mut packed_output_stream = PackedOutputStream {inner : output};
-    serialize::write_message(&mut packed_output_stream, message)
-}
-
-
-pub fn write_packed_message_unbuffered<T: OutputStream, U: MessageBuilder>(
-    output : &mut T, message : &mut U) -> ::std::io::Result<()> {
-    let mut buffered = BufferedOutputStreamWrapper::new(output);
-    try!(write_packed_message(&mut buffered, message));
-    buffered.flush()
+pub fn write_message<W, M>(write: &mut W, message : &mut M) -> io::Result<()>
+where W: Write, M: MessageBuilder {
+    let mut packed_write = PackedWrite { inner: write };
+    serialize::write_message(&mut packed_write, message)
 }
 
 #[cfg(test)]
 mod tests {
-    use std;
-    use serialize_packed::{PackedOutputStream, PackedInputStream};
-    use io::{ArrayInputStream, ArrayOutputStream, InputStream, OutputStream};
+
+    use std::iter;
+    use std::io::Write;
+
+    use std::io::Cursor;
+    use quickcheck::{quickcheck, TestResult};
+
+    use {Word, MessageReader};
+    use message::ReaderOptions;
+    use serialize::test::write_message_segments;
+    use serialize_packed::{PackedRead, PackedWrite};
+    use super::read_message;
+    use util::read_exact;
+
+
```

Comment:
I haven't actually changed this- let me know if you want me to take the spaces out.  I introduced this style to quite a few places.